Use ginkgo's new Skip()

karlkfi opened the pull request:
The following kinds of early returns or failures have been replaced with skips:
- not supported by the current platform
- requires at least 2 nodes

Ginkgo now verbosely prints out what tests were skipped at runtime.

This will allow reduced capability providers (like vagrant or local docker) to skip tests for functionality they don't plan to support. 

Long term, we should try to make it more clear while features are required for minimal compatibility, and which are added by specific providers.

Commit:
Update Ginkgo

- Adds support for Skip()

Commit:
Use Ginkgo Skip() to avoid failing when tests are skipped

- Added util methods: Skipf, SkipUnlessNodeCountIsAtLeast, SkipIfProviderIs,
  and SkipUnlessProviderIs

karlkfi commented:
Blocks #9681

roberthbailey commented:
ok to test

roberthbailey commented:
Assigning to @ixdy. /cc @quinton-hoole @mbforbes @jlowdermilk

karlkfi commented:
Example verbose output of a test skipped at runtime:

```
$ export KUBERNETES_PROVIDER=vagrant
$ export NUM_MINIONS=2
$ ./cluster/kube-up.sh
...
$ go run ./hack/e2e.go -v -test -check_version_skew=false '--test_args=--ginkgo.focus=Networking.*should.function.for.intra-pod.communication'
2015/06/22 15:28:12 Running: get status
Vagrant doesn't need special preparations for e2e tests
Client Version: version.Info{Major:"0", Minor:"19+", GitVersion:"v0.19.3-739-g4317bd7464ce09-dirty", GitCommit:"4317bd7464ce0911b504f2d4e03ba7665934e046", GitTreeState:"dirty"}
Server Version: version.Info{Major:"0", Minor:"19+", GitVersion:"v0.19.1-696-gcdd0cf8a8083fe", GitCommit:"cdd0cf8a8083fe8ce2089baf4878046de7011727", GitTreeState:"clean"}
Vagrant doesn't need special preparations for e2e tests
2015/06/22 15:28:13 Running: Ginkgo tests
Setting up for KUBERNETES_PROVIDER="vagrant".
Vagrant doesn't need special preparations for e2e tests
KUBE_MASTER_IP: 10.245.1.2
>>> testContext.KubeConfig: /Users/karl/.kube/config
INFO: Waiting up to 10m0s for all pods (need at least 0) in namespace 'default' to be running and ready
INFO: 1 / 1 pods in namespace 'default' are running and ready (5 seconds elapsed)
Running Suite: Kubernetes e2e suite
===================================
Random Seed: 1435012093 - Will randomize all specs
Will run 1 of 89 specs

SSSSSSSSSSS
------------------------------
Networking
  should function for intra-pod communication
  /Users/karl/workspace/k8s-master/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/networking.go:210
[BeforeEach] Networking
  /Users/karl/workspace/k8s-master/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/framework.go:47
STEP: Creating a kubernetes client
>>> testContext.KubeConfig: /Users/karl/.kube/config
STEP: Building a namespace api object
INFO: Waiting up to 2m0s for service account default to be provisioned in ns e2e-tests-nettest-s25o3
INFO: Get service account default in ns e2e-tests-nettest-s25o3 failed, ignoring for 5s: serviceaccounts "default" not found
INFO: Service account default in ns e2e-tests-nettest-s25o3 found. (5.009651789s)
STEP: Waiting for a default service account to be provisioned in namespace
INFO: Waiting up to 2m0s for service account default to be provisioned in ns e2e-tests-nettest-s25o3
INFO: Service account default in ns e2e-tests-nettest-s25o3 found. (2.336314ms)
[BeforeEach] Networking
  /Users/karl/workspace/k8s-master/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/networking.go:51
STEP: Executing a successful http request from the external internet
[It] should function for intra-pod communication
  /Users/karl/workspace/k8s-master/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/networking.go:210
[AfterEach] Networking
  /Users/karl/workspace/k8s-master/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/framework.go:48
STEP: Destroying namespace "e2e-tests-nettest-s25o3" for this suite.

S [SKIPPING] [5.330 seconds]
Networking
/Users/karl/workspace/k8s-master/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/networking.go:212
  should function for intra-pod communication [It]
  /Users/karl/workspace/k8s-master/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/networking.go:210

  Skipping test which is broken for vagrant (See https://github.com/GoogleCloudPlatform/kubernetes/issues/3580)

  /Users/karl/workspace/k8s-master/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/networking.go:77
------------------------------
SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS
Ran 0 of 89 Specs in 5.331 seconds
SUCCESS! -- 0 Passed | 0 Failed | 0 Pending | 89 Skipped PASS

Ginkgo ran 1 suite in 10.786970547s
Test Suite Passed
```

ixdy commented on test/e2e/certs.go:
```diff
@@ -25,18 +25,19 @@ import (
-			By(fmt.Sprintf("Skipping MasterCerts test for cloud provider %s (only supported for gce and gke)", testContext.Provider))
-			return
+		if !providerIs(supportedProviders...) {
+			Skipf("Only supported for providers %v (not %s)", supportedProviders, testContext.Provider)
```
this pattern is common enough that I wonder if we should have a wrapper which calls `providerIs()` followed by `Skipf` if needed.

ixdy commented on test/e2e/pd.go:
```diff
@@ -122,14 +117,6 @@ var _ = Describe("Pod Disks", func() {
 	})
 
 	It("should schedule a pod w/ a readonly PD on two hosts, then remove both.", func() {
-		if testContext.Provider != "gce" {
```
this one required gce, not gce or aws.

karlkfi replied:
Fixed.

ixdy commented on test/e2e/resize_nodes.go:
```diff
@@ -469,68 +460,67 @@ var _ = Describe("Nodes", func() {
+					Skipf("Only supported for providers %v (not %s)", supportedProviders, testContext.Provider)
+				}
+				if testContext.CloudConfig.NumNodes < 2 {
+					Skipf("Requires at lease 2 nodes (not %d)", testContext.CloudConfig.NumNodes)
```
least.

it seems like the node number check is also a common pattern that may be worth turning into a utility function.

ixdy commented on test/e2e/resize_nodes.go:
```diff
@@ -469,68 +460,67 @@ var _ = Describe("Nodes", func() {
+			// 1. pods from a uncontactable nodes are rescheduled
+			// 2. when a node joins the cluster, it can host new pods.
+			// Factor out the cases into two separate tests.
+			It("[replication controller] recreates pods scheduled on the unreachable minion node "+
```
was changing the name of the test intended here?

karlkfi replied:
Intentional, yes. Tho perhaps not exactly related to the goal of this PR.
The previous name was ambiguous so I changed it to something that made more sense to me. The comment above it indicates that the test needs to be refactored anyway. So it'll probably get renamed shortly.

karlkfi commented:
- Added 3 util methods: `SkipUnlessNodeCountIsAtLeast`, `SkipIfProviderIs`, and `SkipUnlessProviderIs`
- Found and updated a few more tests that were using `providerIs`.
- Fixed the one case that required gce and not aws.

pmorie commented:
DO WANT Skip()

On Mon, Jun 22, 2015 at 10:25 PM, Kubernetes Bot <notifications@github.com>
wrote:

> GCE e2e build/test *passed* for commit c435d81
> <https://github.com/GoogleCloudPlatform/kubernetes/commit/c435d81aff7547edb29d5a5f91de894ce0b23f58>
> .
>
>    - Build Log
>    <https://storage.cloud.google.com/kubernetes-jenkins/pr-logs/c435d81aff7547edb29d5a5f91de894ce0b23f58/kubernetes-pull-build-test-e2e-gce/832/build-log.txt>
>    - Test Artifacts
>    <https://console.developers.google.com/storage/browser/kubernetes-jenkins/pr-logs/c435d81aff7547edb29d5a5f91de894ce0b23f58/kubernetes-pull-build-test-e2e-gce/832/_artifacts/>
>    - Internal Jenkins Results
>    <http://goto.google.com/prkubekins/job/kubernetes-pull-build-test-e2e-gce//832>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/10197#issuecomment-114331603>
> .
>

karlkfi commented:
All comments have been addressed. How's it look, @ixdy ?

ixdy commented:
Sorry for the delay. This generally LG, aside from a few nits - thanks so much for adding this!

ixdy commented:
also, this needs to be rebased.

karlkfi commented:
Rebased. Fixed comment issues.

karlkfi commented:
Rebased. Probably want to run the CI e2e tests again to verify.

quinton-hoole commented:
@k8s-bot Test this please

quinton-hoole commented:
Aargh - Jenkins e2e is backed up.  Forge ahead!  LGTM.  ok-to-merge

quinton-hoole commented:
Jenkins is back up: https://github.com/jenkinsci/ghprb-plugin/issues/117

mbforbes commented:
roping in @roberthbailey for second LGTM

quinton-hoole commented:
Very unfortunate typo above:

Jenkins is BACKED up (i.e. broken/delayed): jenkinsci/ghprb-plugin#117

roberthbailey commented:
LGTM.

FYI @jlowdermilk @mbforbes @zmerlynn @cjcullen One of the test names changed which may require some tweaks to our jenkins configuration for selecting which tests need to be run.

mbforbes commented:
Thanks for your patience sticking with this @karlkfi; just waiting for our CI system(s) to come back. Note to next oncall: please ensure this passes Shippable/Jenkins before merging.

Also, @ixdy, will you need to update #10078 as is implied by https://github.com/GoogleCloudPlatform/kubernetes/pull/10197#issuecomment-115432583 ?

ixdy commented:
@k8s-bot test this please

PR e2e may be recovering. still tracking down the issues.
I will update my PR too.

karlkfi commented:
CI seems happy.

```
Ran 49 of 90 Specs in 290.864 seconds
SUCCESS! -- 49 Passed | 0 Failed | 2 Pending | 39 Skipped 

Ginkgo ran 1 suite in 5m1.164943417s
Test Suite Passed
```

roberthbailey commented:
@mbforbes -- "All is well -- 2 successful checks".
